common/mqtt_client/src/lib.rs

```diff
@@ -184,17 +184,17 @@ impl Client {
                         QoS::ExactlyOnce => {
                             // Do not annouce the incoming publish message immediatly in case
                             // of QoS=2. Wait for the `PubRel`.
-                            //
-                            // TODO: Avoid buffering the message here by
-                            // moving that part of the code into the `rumqttc` library.
                             let _ = pending.insert(msg.pkid, msg.into());
-                            pending_count
-                                .as_ref()
-                                .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
                         }
                     }
                 }
 
+                Ok(Event::Incoming(Packet::PubAck(_))) => {
+                    pending_count
+                        .as_ref()
+                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                }
```

Comment:
Something similar needs to be done for QoS 2 on reception of a [PubComp](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718058) and not of a PubRel that are related to *incoming* messages.

```suggestion
                Ok(Event::Incoming(Packet::PubAck(_))) => {
                    pending_count
                        .as_ref()
                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
                }

                Ok(Event::Incoming(Packet::PubComp(_))) => {
                    pending_count
                        .as_ref()
                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
                }
```